File: Typescript/sortedArrayToBST.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function sortedArrayToBST(nums: number[]): TreeNode | null {
    const sortedArrRecursion = (nums: number[], beg: number, end: number): TreeNode | null => {
        if (beg > end) {
            return null;
        }
        const mid = Math.ceil((beg + end) / 2);
        const root = new TreeNode(nums[mid]);
        root.left = sortedArrRecursion(nums, beg, mid - 1);
        root.right = sortedArrRecursion(nums, mid + 1, end);
        return root;
    }

    return sortedArrRecursion(nums, 0, nums.length - 1);
}

File: C++/sortedArrayToBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(std::vector<int>& nums)
    {
        return sortedArrayToBSTRecursive(nums, 0, nums.size() - 1);
    }

private:
    TreeNode* sortedArrayToBSTRecursive(const std::vector<int>& nums, int left, int right) 
    {
        if (left > right) 
        {
            return nullptr;
        }

        int mid = left + (right - left) / 2;
        TreeNode* root = new TreeNode(nums[mid]);
        root->left = sortedArrayToBSTRecursive(nums, left, mid - 1);
        root->right = sortedArrayToBSTRecursive(nums, mid + 1, right);
        return root;
    }
};